### DISCUSSION TITLE
Big changes for version 3

### DISCUSSION THREAD

**Discussion by andrebaltieri** [andrebaltieri/flunt#112] [Ideas]
## Dictionary instead of List
When we designed Flunt, we intend to store entity validations and warnings nested to its object. This is cool, and we looked forward to let you extend your notifications by creating a `Contract<T>` where T is a type of Notification.

However, the amount of projects that extends the Notification class is very small. In fact is better to keep less data as possible and most case ends in a key/pair (Of string).

The proposal here is to change from `public IReadOnlyList<T> Notifications { get; }` to `public IReadOnlyDictionary<string, string> Errors { get; }`, making usage of `Dictionaries`, which are way faster than generic lists.

### Downsides
You'll be unable to extend notifications with custom attributes and won't be able to register same key twice or more.

## Standalone Contracts
Beside work with notifications for entities, we can use Flunt to validate almost everything. The proposal here is to separete `Contract` from `Notifiable` to increase validation capabilities.

```csharp
// Is possible to create a stand alone contract
var contract = new Contract()
    .IsNullOrEmpty("", "Key", "Value")
    .IsNotNullOrEmpty("", "Key", "Value");
```

```csharp
// Or inherit from Contract
public class Student : Contract
{
    ...
}
```

## Improve base validation
On version 1 we had a contract (interface) `IValidatable` forcing to implement `Validate` method. This was removed on version 2 since force some interface to be implemented is just bad.

The purpose here is to bring it back, but silently, as a virtual method that you can override.
```csharp
public class Student : Contract
{
    public string FirstName { get; set; } = string.Empty;
    public string LastName { get; set; } = string.Empty;
    public string Email { get; set; } = string.Empty;

    public override void Validate()
    {
      // Your validation here
    }
}
```

## Drop AddNotification need
To make simple validations inside a class that inherit from `Contract` we needed to do something like:
```csharp
public override void Validate()
{
    AddNotifications(new Contract<T>()
        .Requires());
}
```

As we merged **Notifiable** and **Contract**, will be easier to validate.
```csharp
public override void Validate()
{
    IsNotNullOrEmpty(FirstName, "FirstName", "Invalid First Name");
    IsNotNullOrEmpty(LastName, "LastName", "Invalid Last Name");
    IsNotNullOrEmpty(Email, "Email", "Invalid E-mail");
}
```

## Drop support for older .NET versions
We are focusing on performance for this version and .NET 7 is a masterpiece. Strings were restructured to increase performance, Regex had a boost, LINQ was enhanced... there are so many good things about it, that Flunt performance will be increase at least 30% just by moving to .NET 7.

### Downsides
.NET versions lower than .NET 7 were not be supported by this version. You can use Flunt V1 and V2, but V3 will be .NET 7 or higher only.

# PLEASE VOTE BELOW
Please, share your thoughts and vote (👍 or 👎) on topics below.

**Comment by andrebaltieri** [comment]
## Use Dictionary instead of List<T>
Do you agree with this?

**Comment by andrebaltieri** [comment]
## Drop support for older .NET versions
Do you agree with this?

**Reply by Gabriel-Moya** [reply]
Well, considering three points:
- v1 and v2 still work
- All software loses support at a certain time
- Someday existing systems will be updated

I think it's valid v3 supports only .net7+

**Reply by renebentes** [reply]
I believe that the LTS version should be considered.
